gaphor/diagram/diagramtoolbox.py

```diff
@@ -91,33 +98,110 @@ def get_tool_def(modeling_language, tool_name):
+    ):
+        assert not diagram_item_has_element(item_class)
 
-    def item_factory(diagram, parent=None):
-        if subject_class:
-            element_factory = diagram.model
-            subject = element_factory.create(subject_class)
-        else:
-            subject = None
+        self._item_class = item_class
+        self._config_func = config_func
 
-        item = diagram.create(item_class, subject=subject)
+    def create_item(self, diagram, parent=None):
+        item = diagram.create(self.item_class(diagram), subject=None)
 
-        if parent and subject and group(parent.subject, item.subject):
+        if self._config_func:
+            self._config_func(item)
+
+        return item
+
+    def item_class(self, diagram):
+        return self._item_class
+
+
+class ElementItemFactory(AbstractItemFactory):
+    def __init__(
+        self, subject_class: Type[Element], config_func: Optional[ConfigFuncType]
+    ):
+        assert has_diagram_item(subject_class)
+
+        self._subject_class = subject_class
+        self._config_func = config_func
+
+    def create_item(self, diagram, parent=None):
+        element_factory = diagram.model
+        subject = element_factory.create(self._subject_class)
+
+        item = diagram.create(self.item_class(diagram), subject=subject)
+
+        if parent and group(parent.subject, item.subject):
             item.change_parent(parent)
 
-        if config_func:
-            config_func(item)
+        if self._config_func:
+            self._config_func(item)
 
         return item
 
-    item_factory.item_class = item_class  # type: ignore[attr-defined]
-    item_factory.subject_class = subject_class  # type: ignore[attr-defined]
-    return item_factory
+    def item_class(self, diagram):
+        return get_diagram_item(self._subject_class, type(diagram))
+
+    def subject_class(self):
+        return self._subject_class
+
+
+class DeferredElementItemFactory(ElementItemFactory):
+    """Creates only item, but depending on the subject class"""
+
+    def create_item(self, diagram, parent=None):
+        item = diagram.create(self.item_class(diagram))
+
+        if self._config_func:
+            self._config_func(item)
+
+        return item
+
+
+def new_element_item_factory(
+    subject_class: Type[Element], config_func: Optional[ConfigFuncType] = None
+):
+    """
+    Creates new element and adds it on the diagram.
+    ``config_func`` may be a function accepting the newly created item.
+    """
+    return ElementItemFactory(subject_class=subject_class, config_func=config_func)
+
+
+def new_deferred_element_item_factory(
+    subject_class: Type[Element], config_func: Optional[ConfigFuncType] = None
+):
+    """
+    Creates new item depending on the diagram, but does not create the element.
+    ``config_func`` may be a function accepting the newly created item.
+    """
+    return DeferredElementItemFactory(
```

Comment:
I'm not sure if "deferred" is the right term. The factory function has no stake in creating the model element.

Reply:
That's true :thinking: 

It is for connectable items which, upon connection create elements or connect existing ones. I'll try to come up with something slightly different.